File: leetcode/1167. Minimum Cost to Connect Sticks/1167.py
# https://helloacm.com/compute-the-minimum-costs-to-connect-the-sticks-using-priority-queue-or-heap/
# https://leetcode.com/problems/minimum-cost-to-connect-sticks/

class Solution:
    def connectSticks(self, sticks: List[int]) -> int:
        heapq.heapify(sticks)
        res = 0
        while len(sticks) > 1:
            a = heapq.heappop(sticks)
            b = heapq.heappop(sticks)
            heapq.heappush(sticks, a + b)
            res += a + b
        return res

File: leetcode/1167. Minimum Cost to Connect Sticks/1167.cpp
// https://helloacm.com/compute-the-minimum-costs-to-connect-the-sticks-using-priority-queue-or-heap/
// https://leetcode.com/problems/minimum-cost-to-connect-sticks/

class Solution {
public:
    int connectSticks(vector<int>& sticks) {
        priority_queue<int , vector<int>, greater<int> pq(begin(sticks), end(sticks)); 
        int res = 0; 
        while (pq.size() > 1) { 
            int first = pq.top();
            pq.pop(); 
            int second = pq.top();
            pq.pop(); 
            res += first + second;
            pq.push(first + second);
        } 
        return res;        
    }
};
